spec/runner_spec.rb

```diff
@@ -0,0 +1,267 @@
+        mock_runners.each do |runner|
+          (1..mock_runner_count).each do |n|
+            text = Faker::RickAndMorty.quote
+            time_travel_to(n.days.ago) { runner.record_successful_post(text) }
+          end
+        end
+
+        mock_runners.each do |runner|
+          expect(runner.posts.size).to eq(mock_runner_count)
+        end
+      end
+
+      after(:each) { ScraperWiki.sqliteexecute('DELETE FROM data') }
+    end
+
+    describe '.run' do
+      let(:url) { Faker::Internet.url('hooks.slack.com') }
+      let(:text) { Faker::Lorem.paragraph(2) }
+
+      before(:each) do
+        set_environment_variable('MORPH_SLACK_CHANNEL_WEBHOOK_URL', url)
+        set_environment_variable('MORPH_LIVE_MODE', 'true')
+      end
+
+      context 'posted in the last fortnight' do
+        it 'does not run the scraper' do
+          # Fake a successful post
+          Jacaranda::BaseRunner.record_successful_post(text)
+          # Then run the runner
+          expect(Jacaranda::BaseRunner.run).to be false
+        end
+      end
+
+      context 'not posted in the last fortnight' do
+        it 'runs the scraper' do
+          VCR.use_cassette('post_to_slack_webhook', match_requests_on: [:host]) do
+            expect(Jacaranda::BaseRunner.run).to be true
+            expect(a_request(:post, url)).to have_been_made.times(1)
+          end
+        end
+      end
+
+      after(:each) do
+        restore_env
+        ScraperWiki.sqliteexecute('DELETE FROM data')
+      end
+    end
+
+    describe '#post_to_slack' do
+      let(:url) { Faker::Internet.url('hooks.slack.com') }
+      let(:text) { Faker::Lorem.paragraph(2) }
+
+      before(:each) { set_environment_variable('MORPH_SLACK_CHANNEL_WEBHOOK_URL', url) }
+
+      context 'posting to Slack is successful' do
+        it 'POSTs to webhook URL' do
+          VCR.use_cassette('post_to_slack_webhook', match_requests_on: [:host]) do
+            Jacaranda::BaseRunner.post(text)
+            expect(a_request(:post, url)).to have_been_made.times(1)
+          end
+        end
+
+        it 'records the message' do
+          VCR.use_cassette('post_to_slack_webhook', match_requests_on: [:host]) do
+            Jacaranda::BaseRunner.post(text)
+            expect(Jacaranda::BaseRunner.posted_in_last_fortnight?).to be true
+          end
+        end
+      end
+
+      context 'posting to Slack is not successful' do
+        it 'does not record the message' do
+          VCR.use_cassette('post_to_slack_webhook_but_fails', match_requests_on: [:host]) do
+            Jacaranda::BaseRunner.post(text)
+            expect(Jacaranda::BaseRunner.posted_in_last_fortnight?).to be false
+          end
+        end
+      end
+
+      after(:each) do
+        restore_env
+      end
+    end
+  end
+
+  describe '.runners' do
+    context 'when filtering with cli option' do
+      it 'returns everything by default' do
+        Jacaranda.parse([])
+        runners = Jacaranda.runners & mock_runners
```

Comment:
I do find these tests a bit confusing. It seems like `.parse` is effectively being tested throughout here as well?

It's confusing that we're testing .runners but then using the `&` method to pick out the common elements in these two arrays, manipulating the result of the execution before we verify them—which makes it unclear what `.runners` is actually expected to do.

Reply:
Yeah, so the `&` is to work around a nasty bug in the tests. 

The tests create mock runner classes, but don't seem to destroy them after each scenario. 

I'll do some more digging to see if I can fix this.

Reply:
> Yeah, so the & is to work around a nasty bug in the tests.

Cool, that definitely sounds like something we want to fix before merging 👍 Feel free to bring me into the nitty gritty if it helps :)